File: source/vire/utility/base_event.cc
// vire/utility/base_event.cc
//
// Copyright (c) 2016 by François Mauger <mauger@lpccaen.in2p3.fr>
//
// This file is part of Vire.
//
// Vire is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Vire is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License

// Ourselves:
#include <vire/utility/base_event.h>

// Third party:
// - BxJsontools:
#include <bayeux/jsontools/boost_type_converters.h>
// - BxProtobuftools:
#include <bayeux/protobuftools/boost_datetime_converters.h>
#include <bayeux/protobuftools/protobuf_factory.h>

// Declare a protobuf registrar instance for the message class:
#include <vire/base_object_protobuf.h>
#include "vire/utility/BaseEvent.pb.h"

BXPROTOBUFTOOLS_REGISTER_CLASS("vire::utility::base_event",
                               vire::utility::BaseEvent)

namespace vire {

  namespace utility {

    base_event::base_event()
    {
      return;
    }

    base_event::base_event(const boost::posix_time::ptime & t_)
    {
      set_timestamp(t_);
      return;
    }

    base_event::~base_event()
    {
      return;
    }

    // virtual
    base_payload::payload_category base_event::get_category() const
    {
      return CATEGORY_EVENT;
    }

    bool base_event::has_timestamp() const
    {
      return vire::time::is_valid(_timestamp_);
    }

    void base_event::set_timestamp(const boost::posix_time::ptime & t_)
    {
      _timestamp_ = t_;
      return;
    }

    void base_event::reset_timestamp()
    {
      vire::time::invalidate_time(_timestamp_);
      return;
    }

    const boost::posix_time::ptime & base_event::get_timestamp() const
    {
      return _timestamp_;
    }

    void base_event::jsonize(jsontools::node & node_,
                             const unsigned long int version_)
    {
      this->base_payload::jsonize(node_, version_);
      node_["timestamp"] % _timestamp_;
      return;
    }

    void base_event::protobufize(protobuftools::message_node & node_,
                                 const unsigned long int version_)
    {
      VIRE_PROTOBUFIZE_PROTOBUFABLE_BASE_OBJECT(base_payload,node_);
      node_["timestamp"] % _timestamp_;
      return;
    }

    void base_event::tree_dump(std::ostream & out_,
                                 const std::string & title_,
                                 const std::string & indent_,
                                 bool inherit_) const
    {
      this->base_payload::tree_dump(out_, title_, indent_, true);

      out_ << indent_ << ::datatools::i_tree_dumpable::inherit_tag(inherit_)
           << "Timestamp : ";
      if (has_timestamp()) {
        out_ << '[' << vire::time::to_string(_timestamp_) << ']';
      } else {
        out_ << "<none>";
      }
      out_ << std::endl;

      return;
    }

  } // namespace utility

} // namespace vire

File: source/java/vire/utility/base_event.java
package vire.utility;

import java.io.*;
import vire.utility.base_payload;
import java.time.Instant;

/// \brief Base event payload
public class base_event
    extends base_payload
{

    private java.time.Instant _timestamp_; //!< Event timestamp	 (MIN is considered as invalid)

    public base_event()
    {
	// Default: MIN is considered as invalid
	_timestamp_ = java.time.Instant.MIN;
	return;
    }

    public base_event(java.time.Instant ts_)
    {
	set_timestamp(ts_);
	return;
    }

    @Override
    public String get_type_id()
    {
	return "vire::utility::base_event";
    }

    @Override
    public String get_protobuf_type_id()
    {
	return "vire.utility.BaseEvent";
    }

    @Override
    public final payload_category get_category()
    {
	return payload_category.CATEGORY_EVENT;
    }

    public boolean has_timestamp()
    {
	return _timestamp_ != java.time.Instant.MIN;
    }

    public void set_timestamp(java.time.Instant ts_)
    {
	_timestamp_ = ts_;
	return;
    }

    public java.time.Instant get_timestamp()
    {
	return _timestamp_;
    }

    public void reset_timestamp()
    {
	_timestamp_ = java.time.Instant.MIN;
	return;
    }

    @Override
    public String toString()
    {
	String s = '[' + get_timestamp().toString() +']';
	return s;
    }

    @Override
    public void tree_dump(OutputStream out_,
			  String title_,
			  String indent_,
			  boolean inherits_)
    {
	super.tree_dump(out_, title_, indent_, true);

	PrintStream out = new PrintStream(out_);

	out.print(indent_ + vire.tree_dumpable_interface.inherit_tag(inherits_)
		  + "Timestamp : ");
	if (! has_timestamp()) {
	    out.print("<none>");
	} else {
	    out.print("[" + get_timestamp() + ']');
	}
	out.println();

	return;
    }

    public static void main(String[] args)
    {
	java.time.Instant timestamp = java.time.Instant.now();
	timestamp.plusNanos(123456789);
	base_event e = new base_event(timestamp);
	if (e.has_timestamp()) {
	    System.out.printf("Base event : %s%n", e);
	}
	e.tree_dump(System.out, "Base event: ", "", false);
	return;
    }

}
